feat: Schedule a follow up segment event for first time passed learners

muhammad-ammar opened the pull request:
<!--

🌰🌰
🌰🌰🌰🌰         🌰 Note: the Nutmeg master branch has been created.  Please consider whether your change
    🌰🌰🌰🌰     should also be applied to Nutmeg. If so, make another pull request against the
🌰🌰🌰🌰         open-release/nutmeg.master branch, or ping @nedbat for help or questions.
🌰🌰

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/edx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

## Description

Describe what this pull request changes, and why. Include implications for people using this change.
Design decisions and their rationales should be documented in the repo (docstring / ADR), per
[OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be
linked here.

Useful information to include:
- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
"Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these
changes.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.

muhammad-ammar committed:
feat: fire follow up event for first time passed learners

jansenk commented on lms/djangoapps/grades/models.py:
```diff
@@ -797,3 +798,27 @@ def _prepare_override_params(subsection_grade_model, override_data):
         return cleaned_data
+
+
+class LearnerCourseEvent(TimeStampedModel):
```
My only issue is that this doesn't really feel like it belongs in the grades app. I like that it's written to be extensible, but in doing so you've written something that doesn't really fit in this box, ya know?

jansenk replied:
Part of me says we should do what compute_grades does with ComputeGradesSetting (having an explicit model defined only for this case) and part of me thinks we should elevate this to some sort of course-level utility rather than burying it in the grades app. 

I think that we should do one of them and not leave it as is, but I'm not really sure which it should be.

muhammad-ammar replied:
Removed the model

jansenk commented on lms/djangoapps/grades/events.py:
```diff
@@ -30,7 +33,7 @@
 COURSE_GRADE_NOW_PASSED_EVENT_TYPE = 'edx.course.grade.now_passed'
 COURSE_GRADE_NOW_FAILED_EVENT_TYPE = 'edx.course.grade.now_failed'
-LEARNER_PASSED_COURSE_FIRST_TIME = 'edx.course.learner.passed.first_time'
+LEARNER_PASSED_COURSE_FIRST_TIME = LEARNER_PASSED_COURSE_FIRST_TIME_EVENT_TYPE
```
I know this is not new, but I don't like the semi collision here between "LEARNER_PASSED_COURSE_FIRST_TIME_EVENT_TYPE" and "COURSE_GRADE_PASSED_FIRST_TIME_EVENT_TYPE". The names imply a difference that isn't really there - they're ultimately the same signal, just one is `tracker.emit(`-ed and one is explicitly emitted to segment. Could this name change to reflect that? `SEGMENT_COURSE_GRADE_PASSED_FIRST_TIME_EVENT_TYPE` or something like that?

jansenk replied:
Also, either move the definitions from the constants file to this file or delete this "rename". You use this "constant" in this file but in all other files you seem to be referencing the value from `constants.py` . Just have one actual constant

jansenk commented on lms/djangoapps/grades/management/commands/send_follow_up_segment_events_for_passed_learners.py:
```diff
@@ -0,0 +1,78 @@
+        follow_up_event_ids = []
+        log.info(f'{log_prefix} Command started.')
+
+        today = timezone.now().date()
```
Is there any reason we'd want to keep these records? Could we delete old ones somewhere? Or mark them as "done" somehow?

irfanuddinahmad commented on lms/djangoapps/grades/config/__init__.py:
```diff
@@ -25,4 +25,5 @@ def should_persist_grades(course_key):
     """
     Returns whether grades should be persisted.
     """
+    return True
```
Is this intentional?

muhammad-ammar replied:
Great catch. Leftover due to debugging.

irfanuddinahmad commented on lms/djangoapps/grades/events.py:
```diff
@@ -215,7 +215,7 @@ def fire_segment_event_on_course_grade_passed_first_time(user_id, course_locator
     * Event should be only fired for learners enrolled in paid enrollment modes.
     """
-    event_name = LEARNER_PASSED_COURSE_FIRST_TIME
+    event_name = 'edx.course.learner.passed.first_time'
```
any particular reason for changing this?

muhammad-ammar replied:
cleanup

irfanuddinahmad reviewed (approved):
Ok when green

jansenk reviewed (approved):
Looks good to me. Just out of curiosity, would you mind tagging me in the PR so I can see what your eventual approach is in the other app?

muhammad-ammar commented:
@jansenk Thank you for the review. The approach is same as we have implemented earlier in grades app. We have created a [new plugin](https://github.com/edx/outcome-surveys) and moved all the implementation there.
